### PR TITLE
Negative `ignore_index` for the Accuracy metric

### PR DESCRIPTION
**Pull request by yassersouri**
As with the `CrossEntropy` loss in Pytorch, it is usual to have negative value for `ignore_index`.
Currently the check in `_check_classifiction_inputs` (specifically `_basic_input_validation`) does not allow a negative value for target.

I have currently created some tests to show the problem.
If this is something that you think should be fixed, I can write a fix for it.

My proposal is to add `ignore_index` as an additional input to the `_basic_input_validation` and slightly change the implementation.
Please let me know if this is something that makes sense and you guys want it fixed.

### CODE REVIEW

**Comment by yassersouri** [comment]
> # [Codecov](https://codecov.io/gh/PyTorchLightning/metrics/pull/362?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PyTorchLightning) Report
> > Merging [#362](https://codecov.io/gh/PyTorchLightning/metrics/pull/362?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PyTorchLightning) ([6667e8c](https://github.com/PyTorchLightning/metrics/commit/6667e8cf45c8a29ee022ded24d9da7aa089c0005)) into [master](https://codecov.io/gh/PyTorchLightning/metrics/commit/3de06f91f479283bde29a212a03a3974cd56cce7?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PyTorchLightning) ([3de06f9](https://github.com/PyTorchLightning/metrics/commit/3de06f91f479283bde29a212a03a3974cd56cce7)) will **decrease** coverage by `20.05%`.

How is it possible to decrease the coverage by adding a test?!

**Comment by yassersouri** [comment]
All I need is a thumbs up or thumbs down on this.
Currently for me it makes a lot of sense to have negative value for ignore_index.

**Comment by SkafteNicki** [comment]
@yassersouri fine by me. The changes probably need to be done to the `StatScore` class as the other classification metric inherit from this.

**Comment by Borda** [comment]
@yassersouri @SkafteNicki how is it going here? :rabbit:

**Comment by yassersouri** [comment]
Hi again! Sorry for disappearing for a while. I will work on this pull request again.
There were some issues regarding the way the ignored items are removed from the predictions and targets that won't work in this new scenario. I am thinking about two ways to solve it.
1. Adding a new index (num_classes + 1) for the ignore index and deal with it that way. 
2. Converting the inputs from d-dimensional to 1-dimensional so it is easier to remove the ignored indices.

Currently I think option 1 is messy as it requires making changes at multiple locations. I will try to implement version 2 and then gather your feedback.

**Comment by yassersouri** [comment]
I just resolved the conflicts, please ignore the review request.

**Comment by yassersouri** [comment]
@SkafteNicki PTAL.
I have done a first attempt at implementing this. I want to get your feedback.

The main change is in `functional/classification/stat_scores.py` in the `_stat_scores_update` function.

Known issue:
When everything is ignored. Then it is harder to deal with as there are multiple places in the code that assumes `preds.numel() > 0`. What do you think I should do about this?
It seem that I have broken some other functionality for `precision_recall` metrics as well.

Cleanup remaining:
1. updating the docstrings
2. merging the tests into a single test with `pytest.mark.parameterize`.

Notes:
 - `ignore_index < num_classes`. Previously it was assumed that `0 <= ignore_index < num_classes`.
 - target can have non-negative values. Previously it was assumed that `target.min() >= 0` in `_basic_input_validation`.
 - In case of Multi-label inputs, it is not easy to ignore negative indices. I don't think the code works in that case, not sure how to handle it.

File: tests/classification/test_accuracy.py [hunk 1]
```diff
@@ -360,3 +360,127 @@ def test_same_input(average):
 
     assert torch.allclose(class_res, torch.tensor(sk_res).float())
     assert torch.allclose(func_res, torch.tensor(sk_res).float())
+
+
+def test_negative_ignore_index():
```
Comment (Borda) [inline tests/classification/test_accuracy.py:365]: seem that the tests are almost the same, mind rather parameterized them?

**Comment by Borda** [comment]
@SkafteNicki any update here? :]

**Comment by Borda** [comment]
@SkafteNicki how is it going here? :rabbit:

**Comment by Borda** [comment]
This seems quite abundant; shall we close it?
cc: @PyTorchLightning/core-metrics

**Comment by stancld** [comment]
@Borda @SkafteNicki Could you please have a look at this PR? I tried to fix some minor issues but I'm still not fully familiar with the `torchmetrics` classification package :]

**Review by Borda — APPROVED** [review]
lgtm
